### PR TITLE
hub csr controller

### PR DESCRIPTION
**Pull request by skeeey**
hub csr controller will
- create the clusterrole/clusterrolebinding after csr is approved
- auto approve the accepted spoke cluster cert rotation

### CODE REVIEW

File: pkg/hub/csr/manifests/clusterrolebinding.yaml [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+metadata:
+  name: system:open-cluster-management:spokecluster:{{ .SpokeClusterName }}-csrapproved
+roleRef:
+  apiGroup: rbac.authorization.k8s.io
+  kind: ClusterRole
+  name: system:open-cluster-management:spokecluster:{{ .SpokeClusterName }}-csrapproved
+subjects:
+- kind: Group
+  apiGroup: rbac.authorization.k8s.io
+  name: system:open-cluster-management:spokecluster:{{ .SpokeClusterName }}
```
Comment (pmorie) [inline pkg/hub/csr/manifests/clusterrolebinding.yaml]: nit: newline eof

File: pkg/hub/csr/controller.go [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/klog"
+)
+
+const (
+	manifestDir   = "pkg/hub/csr"
+	subjectPrefix = "system:open-cluster-management:"
+)
+
+type crsController struct {
```
Comment (pmorie) [inline pkg/hub/csr/controller.go]: typo: `crsController`

File: pkg/hub/csr/controller.go [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/klog"
+)
+
+const (
+	manifestDir   = "pkg/hub/csr"
+	subjectPrefix = "system:open-cluster-management:"
+)
+
```
Comment (pmorie) [inline pkg/hub/csr/controller.go:26]: add a brief type godoc that explains the responsibility of this controller

File: pkg/hub/csr/controller.go [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+	if err != nil {
+		return err
+	}
+
+	approved, denied := getCertApprovalCondition(&csr.Status)
+	if denied {
+		return nil
+	}
+
+	// Recognize whether current csr is a spoker clsuter csr.
```
Comment (pmorie) [inline pkg/hub/csr/controller.go]: typo: cluster

File: pkg/hub/csr/controller_test.go [hunk 5]
```diff
@@ -0,0 +1,299 @@
...
+			csr:        newCSR("", []string{"test"}, "", "CERTIFICATE REQUEST"),
+			recognized: false,
+		},
+		{
+			name:       "an invalid common name",
+			csr:        newCSR("", []string{"system:open-cluster-management:spokecluster1"}, "", "CERTIFICATE REQUEST"),
+			recognized: false,
+		},
+		{
+			name:                "a recognied csr",
```
Comment (pmorie) [inline pkg/hub/csr/controller_test.go]: typo: recognized

File: pkg/hub/csr/controller_test.go [hunk 6]
```diff
@@ -0,0 +1,299 @@
...
+			name:       "an invalid organization",
+			csr:        newCSR("", []string{"test"}, "", "CERTIFICATE REQUEST"),
+			recognized: false,
+		},
+		{
+			name:       "an invalid common name",
+			csr:        newCSR("", []string{"system:open-cluster-management:spokecluster1"}, "", "CERTIFICATE REQUEST"),
+			recognized: false,
+		},
+		{
```
Comment (pmorie) [inline pkg/hub/csr/controller_test.go:157]: I think there's a scenario missing here (which is relied upon in `TestSync`) which is a valid common name that does not match the username.

File: pkg/hub/csr/controller_test.go [hunk 7]
```diff
@@ -0,0 +1,299 @@
...
+					t.Errorf("expected 1 call but got: %#v", actions)
+				}
+				assertAction(t, actions[0], "get")
+			},
+		},
+		{
+			name:         "sync a hub cluster admin approved csr",
+			startingCSRs: []runtime.Object{newApprovedCSR()},
+			validateActions: func(t *testing.T, actions []clienttesting.Action) {
+				if len(actions) != 5 {
```
Comment (pmorie) [inline pkg/hub/csr/controller_test.go]: I really prefer to have the details of each fake call tested

File: pkg/hub/csr/controller.go [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+	recognized, spokeClusterName := recognize(csr)
+	if !recognized {
+		klog.V(4).Infof("CSR %q was not recognized", csr.Name)
+		return nil
+	}
+
+	// Current spoker clsuter csr is approved, apply the corresponding clusterrole/clusterrolebinding.
```
Comment (pmorie) [inline pkg/hub/csr/controller.go]: typo: cluster

**Comment by pmorie** [comment]
/approve

I would be fine with remaining feedback being addressed in a follow-up PR.

File: pkg/hub/csr/controller.go [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+	organization := x509cr.Subject.Organization[0]
+	if !strings.HasPrefix(organization, subjectPrefix) {
+		return false, ""
+	}
+
+	if !strings.HasPrefix(x509cr.Subject.CommonName, organization) {
+		return false, ""
+	}
+
+	return csr.Spec.Username == x509cr.Subject.CommonName, strings.TrimPrefix(organization, subjectPrefix)
```
Comment (deads2k) [inline pkg/hub/csr/controller.go]: make sure the organization matches exactly as well.

File: pkg/hub/csr/controller.go [hunk 10]
```diff
@@ -0,0 +1,195 @@
...
+		return false, err
+	}
+	return sar.Status.Allowed, nil
+}
+
+// To recognize a valid spoke cluster csr, we check
+// 1. if organization field and commonName field in csr request is valid.
+// 2. if user name in csr is the same as commonName field in csr request.
+func recognize(csr *certificatesv1beta1.CertificateSigningRequest) (recognized bool, spokeClusterName string) {
+	block, _ := pem.Decode(csr.Spec.Request)
```
Comment (deads2k) [inline pkg/hub/csr/controller.go:138]: make sure that this is for the expected signer.

File: pkg/hub/csr/controller.go [hunk 11]
```diff
@@ -0,0 +1,195 @@
...
+	if err != nil {
+		return false, err
+	}
+	return sar.Status.Allowed, nil
+}
+
+// To recognize a valid spoke cluster csr, we check
+// 1. if organization field and commonName field in csr request is valid.
+// 2. if user name in csr is the same as commonName field in csr request.
+func recognize(csr *certificatesv1beta1.CertificateSigningRequest) (recognized bool, spokeClusterName string) {
```
Comment (deads2k) [inline pkg/hub/csr/controller.go]: This is actually `isSpokeClusterClientCertRenewal`

Comment (deads2k) [inline pkg/hub/csr/controller.go]: Add a max duration for this CSR.  I think 30 days is long enough to start.

File: pkg/hub/csr/manifests/clusterrole.yaml [hunk 12]
```diff
@@ -0,0 +1,13 @@
...
+  name: system:open-cluster-management:spokecluster:{{ .SpokeClusterName }}-csrapproved
+rules:
+# Allow spoke agent to create/get/list/watch certificatesigningrequests
+- apiGroups: ["certificates.k8s.io"]
+  resources: ["certificatesigningrequests"]
+  verbs: ["create", "get", "list", "watch"]
+# Allow spoke agent to create/get/list/watch spoke cluster
+- apiGroups: ["cluster.open-cluster-management.io"]
+  resources: ["spokeclusters"]
+  verbs: ["create", "get", "list", "watch"]
```
Comment (deads2k) [inline pkg/hub/csr/manifests/clusterrole.yaml]: a spoke should not have create powers.

File: pkg/hub/csr/manifests/clusterrole.yaml [hunk 13]
```diff
@@ -0,0 +1,13 @@
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: system:open-cluster-management:spokecluster:{{ .SpokeClusterName }}-csrapproved
+rules:
+# Allow spoke agent to create/get/list/watch certificatesigningrequests
+- apiGroups: ["certificates.k8s.io"]
+  resources: ["certificatesigningrequests"]
+  verbs: ["create", "get", "list", "watch"]
```
Comment (deads2k) [inline pkg/hub/csr/manifests/clusterrole.yaml]: These abilities should be granted in the primary clusterrole for spokes.

File: pkg/hub/csr/controller.go [hunk 14]
```diff
@@ -0,0 +1,195 @@
...
+			syncCtx.Recorder(),
+			func(name string) ([]byte, error) {
+				config := struct {
+					SpokeClusterName string
+				}{
+					SpokeClusterName: spokeClusterName,
+				}
+				return assets.MustCreateAssetFromTemplate(name, bindata.MustAsset(filepath.Join(manifestDir, name)), config).Data, nil
+			},
+			"manifests/clusterrole.yaml",
```
Comment (deads2k) [inline pkg/hub/csr/controller.go]: I don't see a reason for RBAC resources to be conditional on a CSR approval.  I think the RBAC resource can be created unconditionally and assigned to the user like the previous PR.

Comment (deads2k) [reply pkg/hub/csr/controller.go]: That would eliminate this logic block entirely.

File: pkg/hub/csr/controller.go [hunk 15]
```diff
@@ -0,0 +1,195 @@
...
+	})
+	_, err = c.kubeClient.CertificatesV1beta1().CertificateSigningRequests().UpdateApproval(ctx, csr, metav1.UpdateOptions{})
+	if err != nil {
+		return err
+	}
+	c.eventRecorder.Eventf("SpokeClusterCSRAutoApproved", "spoke cluster csr %q is auto approved by hub csr controller", csr.Name)
+	return nil
+}
+
+func (c *csrController) authorize(ctx context.Context, csr *certificatesv1beta1.CertificateSigningRequest) (bool, error) {
```
Comment (deads2k) [inline pkg/hub/csr/controller.go]: godoc for future you.

File: pkg/hub/csr/controller.go [hunk 16]
```diff
@@ -0,0 +1,195 @@
...
+	}
+
+	sar := &authorizationv1.SubjectAccessReview{
+		Spec: authorizationv1.SubjectAccessReviewSpec{
+			User:   csr.Spec.Username,
+			UID:    csr.Spec.UID,
+			Groups: csr.Spec.Groups,
+			Extra:  extra,
+			ResourceAttributes: &authorizationv1.ResourceAttributes{
+				Group:       "certificates.k8s.io",
```
Comment (deads2k) [inline pkg/hub/csr/controller.go]: don't use resources in someone else's group.  Create this in your own group.

How about verb=renew, group=register.open-cluster-management.io resource=spokeclusters subresource=clientcertificates name=spokecluster

File: pkg/hub/csr/controller.go [hunk 17]
```diff
@@ -0,0 +1,195 @@
...
+	}
+
+	if !strings.HasPrefix(x509cr.Subject.CommonName, organization) {
+		return false, ""
+	}
+
+	return csr.Spec.Username == x509cr.Subject.CommonName, strings.TrimPrefix(organization, subjectPrefix)
+}
+
+func getCertApprovalCondition(status *certificatesv1beta1.CertificateSigningRequestStatus) (approved, denied bool) {
```
Comment (deads2k) [inline pkg/hub/csr/controller.go]: I think this should be `isCSRInTerminalState`.  if it's been approved or denied before, you have no work to do.

**Comment by deads2k** [comment]
> /approve
> 
> I would be fine with remaining feedback being addressed in a follow-up PR.

The intent is good and the code is easy to read (nicely done), but because this is security related, I'd like to get it safe before we merge.  My comments look bigger than they are, I think it will be easy to stitch in the changes.

**Comment by pmorie** [comment]
> The intent is good and the code is easy to read (nicely done), but because this is security related, I'd like to get it safe before we merge. My comments look bigger than they are, I think it will be easy to stitch in the changes.

Cool, lgtm is on you @deads2k

File: pkg/hub/csr/manifests/clusterrole.yaml (hunk 12 above)
Comment (deads2k) [reply pkg/hub/csr/manifests/clusterrole.yaml]: > a spoke should not have create powers.

I expected create to be the domain of the bootstrap ID and spokes would have named get,list,watch,update,update/status powers

File: pkg/hub/csr/controller.go (hunk 3 above)
Comment (skeeey) [reply pkg/hub/csr/controller.go:26]: added

File: pkg/hub/csr/controller.go (hunk 4 above)
Comment (skeeey) [reply pkg/hub/csr/controller.go]: corrected

File: pkg/hub/csr/controller.go (hunk 8 above)
Comment (skeeey) [reply pkg/hub/csr/controller.go]: corrected

File: pkg/hub/csr/controller.go (hunk 14 above)
Comment (skeeey) [reply pkg/hub/csr/controller.go]: remove this logic entirely

File: pkg/hub/csr/controller.go (hunk 15 above)
Comment (skeeey) [reply pkg/hub/csr/controller.go]: added

File: pkg/hub/csr/controller.go (hunk 16 above)
Comment (skeeey) [reply pkg/hub/csr/controller.go]: +1

File: pkg/hub/csr/controller.go (hunk 11 above)
Comment (skeeey) [reply pkg/hub/csr/controller.go]: rename this function

Comment (skeeey) [reply pkg/hub/csr/controller.go]: What this the max duration, I cannot find it from CSR fields ..., would you help me to point it out?

File: pkg/hub/csr/controller.go (hunk 10 above)
Comment (skeeey) [reply pkg/hub/csr/controller.go:138]: added

File: pkg/hub/csr/controller.go (hunk 9 above)
Comment (skeeey) [reply pkg/hub/csr/controller.go]: I get the spoke cluster name from csr label and use it to match the organization exactly

File: pkg/hub/csr/controller.go (hunk 17 above)
Comment (skeeey) [reply pkg/hub/csr/controller.go]: ok, refactor this

File: pkg/hub/csr/controller_test.go (hunk 7 above)
Comment (skeeey) [reply pkg/hub/csr/controller_test.go]: enhanced

File: pkg/hub/csr/controller_test.go (hunk 6 above)
Comment (skeeey) [reply pkg/hub/csr/controller_test.go:157]: added

File: pkg/hub/csr/controller_test.go (hunk 5 above)
Comment (skeeey) [reply pkg/hub/csr/controller_test.go]: corrected

File: pkg/hub/csr/controller.go [hunk 18]
```diff
@@ -0,0 +1,178 @@
...
+
+	return csr.Spec.Username == x509cr.Subject.CommonName
+}
+
+// Check whether a CSR is in terminal state
+func isCSRInTerminalState(status *certificatesv1beta1.CertificateSigningRequestStatus) bool {
+	approved, denied := false, false
+	for _, c := range status.Conditions {
+		if c.Type == certificatesv1beta1.CertificateApproved {
+			approved = true
```
Comment (deads2k) [inline pkg/hub/csr/controller.go:171]: nit: return true

File: pkg/hub/csr/controller.go [hunk 19]
```diff
@@ -0,0 +1,178 @@
...
+		}
+		if c.Type == certificatesv1beta1.CertificateDenied {
+			denied = true
```
Comment (deads2k) [inline pkg/hub/csr/controller.go:174]: nit: return true

File: pkg/hub/csr/controller.go [hunk 20]
```diff
@@ -0,0 +1,178 @@
...
+		return nil
+	}
+
+	// Authorize whether the current spoke agent has been authorized to renew its csr.
+	allowed, err := c.authorize(ctx, csr)
+	if err != nil {
+		return err
+	}
+	if !allowed {
+		klog.V(4).Infof("Spoke cluster csr %q cannont be auto approved due to subject access review was not approved", csr.Name)
```
Comment (deads2k) [inline pkg/hub/csr/controller.go:77]: TODO find a way to avoid looking at this CSR again.

File: pkg/hub/csr/controller.go [hunk 21]
```diff
@@ -0,0 +1,178 @@
...
+		return nil
+	}
+
+	// Auto approve the spoke cluster csr
+	csr.Status.Conditions = append(csr.Status.Conditions, certificatesv1beta1.CertificateSigningRequestCondition{
+		Type:    certificatesv1beta1.CertificateApproved,
+		Reason:  "AutoApproved",
```
Comment (deads2k) [inline pkg/hub/csr/controller.go:84]: Indicate that *this particular operator* auto approved.

File: pkg/hub/manager.go [hunk 22]
```diff
@@ -23,7 +25,9 @@ func RunControllerManager(ctx context.Context, controllerContext *controllercmd.
 	if err != nil {
 		return err
 	}
+
 	clusterInformers := clusterv1informers.NewSharedInformerFactory(clusterClient, 10*time.Minute)
+	csrInformers := kubeinformers.NewSharedInformerFactory(kubeClient, 10*time.Minute)
```
Comment (deads2k) [inline pkg/hub/manager.go:30]: minor: `kubeInfomers := `  We will use these informers for all kube things.

**Comment by deads2k** [comment]
minor comments, but they can be done post-merge

/lgtm

File: pkg/hub/csr/controller.go (hunk 20 above)
Comment (skeeey) [reply pkg/hub/csr/controller.go:77]: add this todo

File: pkg/hub/csr/controller.go (hunk 21 above)
Comment (skeeey) [reply pkg/hub/csr/controller.go:84]: done

File: pkg/hub/csr/controller.go (hunk 18 above)
Comment (skeeey) [reply pkg/hub/csr/controller.go:171]: +1, clever

File: pkg/hub/csr/controller.go (hunk 19 above)
Comment (skeeey) [reply pkg/hub/csr/controller.go:174]: done

File: pkg/hub/manager.go (hunk 22 above)
Comment (skeeey) [reply pkg/hub/manager.go:30]: done

**Pull request merged**